### ISSUE TITLE
FEATURE REQUEST: Being able to train on N different folds during cross-validation

### ISSUE THREAD

**Issue by egorrevkov** [autonomio/talos#192]
Suggesting a feature following the [discussion](https://spectrum.chat/talos/general/can-talos-train-on-n-different-folds-during-cross-validation~4cba6514-c30f-41f9-95e7-c03cc91d992c) with @mikkokotila on Spectrum:

Currently, the default _ta.Scan(..., val_split, ...)_ behaviour only allows for putting aside the _val_split_  proportion of data out of training for later evaluation of several pre-trained models from the hyperparameter space. For better error estimation, Talos also supports **N-fold cross-evaluation**  via _scan_object.evaluate_models(x_val, y_val, n=N)_ method.

However, it would be great if we could also could automatically **train** our models on N different folds of the training data and then average resulting metrics for proper **cross-validation**.

**Comment by mikkokotila** [comment]
Just to be clear, what you are suggesting is that cross-validation is used as a training metric instead of the Keras way which is to just use whatever metric have been input as `metric` in `model.compile()`.

**Comment by egorrevkov** [comment]
Yes, I am suggesting that cross-validation would be a great optional method for calculating a training metric. But I think the exact `metric` used in for the error calculation (say, mse or logcosh for regression) could still be passed through `model.compile()`.

**Comment by mikkokotila** [comment]
I have been working on this today. I had not done a custom Keras metric before, so needed to learn a couple of things. As it stands, I did not find a way to get the length / number of observations inside the metric for y_pred and y_true which is needed for the split. Otherwise everything is working already. I've posted an issue on Keras and also a question on StackOverflow (as I had already tried ~10 different methods and none returned the actual length of the underlying data for y_pred and y_true). I prefer doing it 100% using Keras backend.

Basically the way this would work is that there are several custom metrics (I think accuracy, f1, and mae at first) which support cross-validation for every epoch.

**Comment by mikkokotila** [comment]
Having thought about this more, it appears tricky. 

Let's  say your batch size is 50 but the whole training dataset is 500 samples. In that case there are 10 batches that will be evaluated for each epoch, and it appears that in fact those batches are kind of like folds. If I understand it correct, for every epoch Keras will pass the batch_size samples to the metric, and the result for each batch will make up the epoch result. In other words, it is already a kind of "fold". 

Again, Keras metrics is something that I have not been working with in the past to this extent, and this is as much as I have been able to gather.

**Comment by cartof** [comment]
Wouldn't be a middle solution to change the train and eval split at each round? Different models will have different train/val pairs but if you have enough models in your grid, some models will be similar and will run on different sets of train/val splits.

Thanks for this great work.

**Comment by mikkokotila** [comment]
@cartof that's an interesting idea! :) This is easy to do just by passing `validation_split` in the params dictionary (or data could be split in some custom way inside the input model also). 

When it comes to a custom metric that would do something like cross-validation as part of the training, I did not find it feasible after spending some time with it. So closing here.
